File: 프로그래머스/lv2/43165. 타겟 넘버/타겟 넘버.py
answer = 0
tmp = []
def dfs(depth, op, numbers, target):
    global answer
    
    if depth == len(numbers):
        if sum(tmp) == target:
            answer += 1
        return
    if op == '+':
        tmp.append(numbers[depth])
    else:
        tmp.append(-numbers[depth])  
    dfs(depth + 1, '+', numbers, target)
    dfs(depth + 1, '-', numbers, target)
    tmp.pop()
    return 
    
def solution(numbers, target):    
    dfs(0, '+', numbers, target)
    tmp = []
    dfs(0, '-', numbers, target)
    tmp = []
    return answer // 2

File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.java
import java.util.*;
class Solution {
    static HashSet<String> nums = new HashSet<>();
    static int answer;
    static void dfs(String s, int cur, int next, int target, int[] numbers){
        if(next == numbers.length){
            if(!nums.contains(s) && cur == target){
                nums.add(s);
                answer++;
            }
            return;
        }
        dfs(s + "-" + numbers[next], cur - numbers[next], next + 1, target, numbers);
        dfs(s + "+" + numbers[next], cur + numbers[next], next + 1, target, numbers);
    }
    public int solution(int[] numbers, int target) {
        dfs("-" + numbers[0], numbers[0], 1, target, numbers);
        dfs("+" + numbers[0], -numbers[0], 1, target, numbers);
        return answer;
    }
}